### PR TITLE
Kp3r allocator v2

### PR DESCRIPTION
**Pull request by 0xLienid**
Still need to figure out how to make rewards work on forked mainnet for the gain and loss focused tests

### CODE REVIEW

File: contracts/allocators/Kp3rAllocatorV2.sol [hunk 1]
```diff
@@ -0,0 +1,228 @@
...
+
+    function _update(uint256 id) internal override returns (uint128 gain, uint128 loss) {
+        if (id == 2) {
+            loss = usdcUsed;
+        } else {
+            uint256 balance = _tokens[0].balanceOf(address(this));
+            (uint256 lockedBalance, ) = kp3rVault.locked(address(this));
+
+            if (balance > 0 && lockedBalance == 0) {
+                lockEnd = block.timestamp + MAX_TIME;
```
[the remark this answers is not in the document]
Comment (PARABRAHMAN0) [reply contracts/allocators/Kp3rAllocatorV2.sol:105]: So, a setter for time. I think that would be ok. It could also just scale below... so when it extends time that it also extends just by the same amount, since after the first lock it will anyways turn into the extend lock value.

**Review by PARABRAHMAN0 — CHANGES_REQUESTED** [review]

File: contracts/allocators/Kp3rAllocatorV2.sol [hunk 2]
```diff
@@ -0,0 +1,228 @@
...
+        _tokens[1].approve(address(rKp3r), type(uint256).max);
+
+        // approve extender
+        _tokens[0].approve(address(treasury), type(uint256).max);
+        _tokens[1].approve(address(treasury), type(uint256).max);
+        rKp3r.approve(address(treasury), type(uint256).max);
+    }
+
+    function _activate() internal override {
+        if (extender.getAllocatorLimits(2).loss < 50000000000) revert Kp3rAllocatorV2_LowUSDCLimits();
```
[the remark this answers is not in the document]
Comment (0xLienid) [reply contracts/allocators/Kp3rAllocatorV2.sol:94]: is $50,000 usdc like we discussed. usdc just has 6 decimals

File: contracts/allocators/Kp3rAllocatorV2.sol [hunk 3]
```diff
@@ -0,0 +1,228 @@
...
+    }
+
+    function _update(uint256 id) internal override returns (uint128 gain, uint128 loss) {
+        if (id == 2) {
+            loss = usdcUsed;
```
[the remark this answers is not in the document]
Comment (0xLienid) [reply contracts/allocators/Kp3rAllocatorV2.sol:99]: good call

File: contracts/allocators/Kp3rAllocatorV2.sol [hunk 4]
```diff
@@ -0,0 +1,228 @@
...
+                _claimAndExercise();
+
+                balance = _tokens[0].balanceOf(address(this));
+
+                if (balance > 0) {
+                    kp3rVault.increase_amount(balance);
+
+                    if (_canExtendLock()) {
+                        lockEnd = block.timestamp + MAX_TIME;
+                        kp3rVault.increase_unlock_time(lockEnd);
```
[the remark this answers is not in the document]
Comment (0xLienid) [reply contracts/allocators/Kp3rAllocatorV2.sol:119]: not sure, up to policy what they want. but the options if not are either a setter for MAX_TIME, which I've already added, and we just set it to 0 before an update, or we can pass in a boolean called extend_ or something. which do you prefer?

File: contracts/allocators/Kp3rAllocatorV2.sol (hunk 1 above)
[the remark this answers is not in the document]
Comment (0xLienid) [reply contracts/allocators/Kp3rAllocatorV2.sol:105]: added an onlyGuardian setter for MAX_TIME

File: contracts/allocators/Kp3rAllocatorV2.sol [hunk 5]
```diff
@@ -0,0 +1,228 @@
...
+            loss = usdcUsed;
+        } else {
+            uint256 balance = _tokens[0].balanceOf(address(this));
+            (uint256 lockedBalance, ) = kp3rVault.locked(address(this));
+
+            if (balance > 0 && lockedBalance == 0) {
+                lockEnd = block.timestamp + MAX_TIME;
+
+                kp3rVault.create_lock(balance, lockEnd);
+            } else if (balance > 0 || lockedBalance > 0) {
```
[the remark this answers is not in the document]
Comment (0xLienid) [reply contracts/allocators/Kp3rAllocatorV2.sol:108]: yeah changed to just lockedBalance > 0 and left the check in 114 since that requires a kp3r balance

File: contracts/allocators/Kp3rAllocatorV2.sol [hunk 6]
```diff
@@ -0,0 +1,228 @@
...
+
+            (lockedBalance, ) = kp3rVault.locked(address(this));
+            uint256 last = extender.getAllocatorAllocated(id) + extender.getAllocatorPerformance(id).gain;
+
+            if (lockedBalance >= last) gain = uint128(lockedBalance - last);
+            else loss = uint128(last - lockedBalance);
+        }
+    }
+
+    function deallocate(uint256[] memory amounts) public override {
```
[the remark this answers is not in the document]
Comment (0xLienid) [reply contracts/allocators/Kp3rAllocatorV2.sol:132]: cool yeah, still need to add comments for everything

File: contracts/allocators/Kp3rAllocatorV2.sol (hunk 2 above)
[the remark this answers is not in the document]
Comment (PARABRAHMAN0) [reply contracts/allocators/Kp3rAllocatorV2.sol:94]: Ok, cool.

File: contracts/allocators/Kp3rAllocatorV2.sol (hunk 4 above)
[the remark this answers is not in the document]
Comment (PARABRAHMAN0) [reply contracts/allocators/Kp3rAllocatorV2.sol:119]: no no, the setter is fine, the only thing that i would do is rename the variable to something else, possibly something like `timeExtension`

**Review by philipjonsen — APPROVED** [review]

**Comment by 0xLienid** [comment]
Rebuilding PR from local fork to try to fix my messed up merge where I dragged hundreds of other commits in
